### ISSUE TITLE
Importing replay and clicking notification crashes the game

### ISSUE THREAD

**Issue by smoogipoo** [ppy/osu#20528]
### Type

Crash to desktop

### Bug description

While at song select, if you import a replay and click the progress notification, the game may crash. It doesn't happen every time but once every 1-3 attempts.

In my logs below, I imported the same replay (https://osu.ppy.sh/scores/taiko/174794219) twice, and it crashed on the second attempt.

### Version

f3184525107b29018d7271e7bbcf779eec7ad1d8

### Logs

[runtime.log](https://github.com/ppy/osu/files/9671412/runtime.log)
[network.log](https://github.com/ppy/osu/files/9671413/network.log)
[database.log](https://github.com/ppy/osu/files/9671415/database.log)
[performance.log](https://github.com/ppy/osu/files/9671416/performance.log)
[input.log](https://github.com/ppy/osu/files/9671417/input.log)
[performance-update.log](https://github.com/ppy/osu/files/9671418/performance-update.log)
[performance-input.log](https://github.com/ppy/osu/files/9671420/performance-input.log)
[performance-draw.log](https://github.com/ppy/osu/files/9671421/performance-draw.log)
[performance-audio.log](https://github.com/ppy/osu/files/9671422/performance-audio.log)

**Comment by smoogipoo** [comment]
Seems like additional steps to repro are:
1. Ensure you have "Show converted" disabled.
2. Select any osu-ruleset difficulty of https://osu.ppy.sh/beatmapsets/205309#osu/488004
3. Import a taiko replay from the same set (https://osu.ppy.sh/scores/taiko/174794219)

**Comment by peppy** [comment]
This is a super hard one to fix. I've tried multiple methods but haven't come up with anything I'm satisfied with. It stems from song select trying to be smart about beatmap and ruleset changes, but having no concept of when an external change requires a specific combination of beatmap and ruleset.

All selection logic at song select ends up going through:

https://github.com/ppy/osu/blob/0815b01b75e2de9364118fc05c6e9410899c8d2a/osu.Game/Screens/Select/SongSelect.cs#L510

even when using `FinaliseSelection` with a specific ruleset provided. This line will cause the beatmap to be filtered away if converts is turned on.

We have complex handling around this code in place to make sure that handling *beatmap* imports can work when already at song select, but it doesn't work for *beatmap+ruleset*.

There's multiple other similar issues with this flow:

- Start at song select in taiko ruleset with "show converted" off
- Import and click replay
- Results screen loads but ruleset is osu!

- Start at song select in taiko ruleset with "show converted" **on**
- Enter any search criteria to cause no matches to display
- Import and click replay
- Crash

Honestly I'm thinking that this flow needs a refactor to give song select the concept of a single "pending selection" tuple `(Beatmap?, Ruleset?)` and hopefully do away with the "decoupled" ruleset nonsense in the process.

This isn't a quick/easy task though. For now, I've opened #20567 to address all the cases I noticed here.
